### PR TITLE
Adds weighted loss to the kgcn learner

### PR DESCRIPTION
**Pull request by TNO-Knowledge-Based-Systems**
## What is the goal of this PR?

The goal of this PR is to be able to use weighted (to the label prevalence) loss for the kgcn learner. This is especially valuable if the labels are not equally distributed.

## What are the changes implemented in this PR?

These changes only affect the two loss functions: loss_ops_preexisting_no_penalty and  loss_ops_from_difference we calculate the label prevalence and calculate a weight from the prevalence using this function: weights = (1 / (label_prevalence + 1) this results in a balanced weighing of the label weights. The default functionality is set to not using weighted loss so this should not impact others.

### CODE REVIEW

**Comment by jmsfltchr** [comment]
I had the thought that this should be something that TensorFlow supports, and it seems that is [somewhat the case](https://github.com/tensorflow/tensorflow/issues/10021), we could look at doing it that as described [here](https://github.com/tensorflow/tensorflow/issues/10021#issuecomment-304253855) and [here](https://github.com/tensorflow/tensorflow/issues/10021#issuecomment-493339191) rather than our own implementation

**Comment by jmsfltchr** [comment]
We have radically changed the repo since this point and deprecated the KGCN, so I'm closing this in favour of https://github.com/vaticle/typedb-ml/pull/161.
